src/rusty/src/rest_downloader.rs

```diff
@@ -0,0 +1,325 @@
+                            chunk_remaining = chunklen;
+                            match lineiter.next() {
+                                Some(l) if l.is_empty() => {
+                                    // Drop \r after \n
+                                    bytes_skipped += 1;
+                                    if actual_len + bytes_skipped > bytes_read {
+                                        // Go back and get more bytes so we can skip trailing \n
+                                        chunk_remaining = 0;
+                                    }
+                                },
+                                Some(_) => {},
+                                None => {
+                                    // Go back and get more bytes so we can skip trailing \n
+                                    chunk_remaining = 0;
+                                },
+                            }
+                            break;
+                        },
+                        Err(_) => return None,
+                    }
+                }
+                if chunk_remaining != 0 {
+                    bytes_read -= bytes_skipped;
+                    resp.drain(actual_len..actual_len + bytes_skipped);
+                    if actual_len + chunk_remaining > max_resp { return None; }
+                    let already_in_chunk = cmp::min(bytes_read - actual_len, chunk_remaining);
+                    actual_len += already_in_chunk;
+                    chunk_remaining -= already_in_chunk;
+                    continue 'read_bytes;
+                } else {
+                    if finished_read {
+                        // Note that we may leave some extra \r\ns to be read, but that's OK,
+                        // we'll ignore then when parsing headers for the next request.
+                        resp.resize(actual_len, 0);
+                        return Some(resp);
+                    } else {
+                        // Need to read more bytes to figure out chunk length
+                    }
+                }
+            }
+            resp.resize(bytes_read + cmp::max(10, chunk_remaining), 0);
+            let avail = read_socket!();
+            bytes_read += avail;
+            if chunk_remaining != 0 {
+                let chunk_read = cmp::min(chunk_remaining, avail);
+                chunk_remaining -= chunk_read;
+                actual_len += chunk_read;
+            }
+        }
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn init_fetch_rest_blocks(uri: *const u8) -> bool {
+    let mut uri_vec = Vec::with_capacity(50);
+    for i in 0..10000 {
+        let c = unsafe { *uri.offset(i) };
+        if c == 0 { break; }
+        uri_vec.push(c);
+    }
+    let uri_str = match String::from_utf8(uri_vec) {
+        Ok(r) => r,
+        Err(_) => return false,
+    };
+    // Sadly only non-SSL is supported for now
+    if let Some((false, _, _, _)) = split_uri(&uri_str) { } else { return false; }
```

Comment:
I think this should be reversed to check if SSL is true.
Because if this returns None we might as well also return false and finish this early, because we unwrap this few lines after 

Better yet, run it once and check for `if ssl {return False;}` so that we won't need to run this twice(not that it really matters that much...)

Reply:
Hmm? The pattern hers is if let Some() {} else { return; }. It reads a bit weird but its the easiest way to write it out.